File: week3_paths_in_graphs1/2_bipartite/bipartite.cpp
#include <iostream>
#include <vector>
#include <queue>

using std::queue;
using std::vector;

bool bfs(int start, vector<vector<int>> &adj, vector<int>&colors)
{
  queue<int> q;
  q.push(start);
  colors[start] = 0;
  
  while (!q.empty())
  {
    int node = q.front();
    q.pop();

    for (int neighbor : adj[node])
    {
      if (colors[neighbor] == -1)
      {
        colors[neighbor] = 1 - colors[node];
        q.push(neighbor);
      }
      else if (colors[neighbor] == colors[node])
      {
        return false;
      }
    }
  }
  return true;
}
int bipartite(vector<vector<int>> &adj)
{
  int n = adj.size();
  vector<int> colors(n, -1);

  for (int i = 0; i < n; i++)
  {
    if (colors[i] == -1)
    {
      if (!bfs(i, adj, colors))
      {
        return 0;
      }
    }
  }
  return 1;
}

int main()
{
  int n, m;
  std::cin >> n >> m;
  vector<vector<int>> adj(n, vector<int>());
  for (int i = 0; i < m; i++)
  {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  std::cout << bipartite(adj);
}

File: week3_paths_in_graphs1/2_bipartite/Bipartite.java
import java.util.*;

public class Bipartite {

    private static boolean bfs(int start, ArrayList<Integer>[] adj, int[] colors) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(start);               // push into queue
        colors[start] = 0;            // assign first color

        while (!q.isEmpty()) {
            int node = q.poll();      // remove from queue

            for (int neighbor : adj[node]) {
                if (colors[neighbor] == -1) {
                    colors[neighbor] = 1 - colors[node];  // opposite color
                    q.offer(neighbor);
                } 
                else if (colors[neighbor] == colors[node]) {
                    return false;     // same color conflict
                }
            }
        }
        return true;
    }

    private static int bipartite(ArrayList<Integer>[] adj) {
        int n = adj.length;
        int[] colors = new int[n];

        Arrays.fill(colors, -1);   

        for (int i = 0; i < n; i++) {
            if (colors[i] == -1) {
                if (!bfs(i, adj, colors)) {
                    return 0;
                }
            }
        }
        return 1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();

        ArrayList<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            int x = scanner.nextInt();
            int y = scanner.nextInt();

            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }

        System.out.println(bipartite(adj));
    }
}